site-src/geps/gep-1282.md

```diff
@@ -98,12 +98,302 @@ Of course, from the section above, you can see that there are many other feature
+
+In https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1beta1/shared_types.go add a new type:
+```go
+// BackendCapabilities references a Kubernetes Service and defines any backend 
+// capabilities that Service requires.  
+//
+// +optional
+type BackendCapabilities struct {
+    // BackendName is the name of the backend service to which these backend 
+    // capabilities apply.
+    // +required
+    BackendName string `json:”backendName”`
+
+    // BackendType is the type of backend to which these backend 
+    // capabilities apply.
+    //
+    // +required
+    BackendType BackendTargetRef `json:”backendType”`
+
+	// Spec specifies the capabilities supported by this
+    // Service wrapper, if any.  The capabilities are predefined, and have two
+    // characteristics in common: they are tightly bound to the backend, 
+    // rather than being important at the route level, and the service owner
+    // (application developer in the Gateway API personas) should control 
+    // these settings, not the owner of the Gateway implementation (the 
+    // Cluster Operator in the Gateway API personas).
+    // 
+    // Support: Extended
+	//
+    // +required
+	Spec BackendCapabilitiesSpec `json:”spec”`
+
+	// Status represents the current state of the backend and its capabilities.
+    // +optional
+	Status *BackendCapabilitiesStatus `json:”status,omitempty”`
+}
+
+// BackendTargetReference identifies an API object to apply BackendCapabilities 
+// to. 
+// 
+type BackendTargetReference struct {
+	// Group is the group of the target resource.
+	Group Group `json:"group"`
+
+	// Kind is kind of the target resource.
+    // Valid values include:
+    // * "Service" (the default if no Kind is specified)
+    // * "ServiceImport"
+    //
+    // Invalid values include:
+    //
+    // * "invalid/kind" - "/" is an invalid character
+    // +required
+	Kind string `json:"kind"`
+
+	// Version is the version of the target resource.
+	Version ObjectName `json:"version"`
+
+	// Name is the name of the target resource.
+	Name ObjectName `json:"name"`
+}
+
+// BackendCapabilitiesSpec defines a group of specific properties held by its
+// affiliated BackendCapabilitiesService, and specified by the Service owner 
+// (typically the application developer).  Examples of properties that 
+// belong in the BackendCapabilitiesSpec include: TLS re-encryption, session
+// state, timeouts, retries, Websockets, HTTP/2, load balancer algorithm, 
+// healthcheck info, external traffic policy, connection limits, error page,
+// protocol switch, alerts, and others. 
+//
+// Support: Extended
+//
+type BackendCapabilitiesSpec struct {
+
+	// +optional
+	Reencrypt *TLSReencryptType `json:”reencrypt,omitempty”`
+
+	// +optional
+    Websocket *WebsocketType `json:”websocket,omitempty”`
+
+	// +optional
+    Protocol *ProtocolType `json:”protocol,omitempty”`
+}
+
+type ReencryptType struct {
+    Version *TLSVersion
+    ClientCertFile string
```

Comment:
Based on the discussion in the community meeting yesterday, this doesn't seem like the right place for this config. Maybe the Gateway is a better fit?

Reply:
I was just commenting in old (closed) issue https://github.com/kubernetes-sigs/gateway-api/issues/622#issuecomment-1267937777 that there might have been some confusion, but since I was not present at the meeting yesterday, I miss the context and maybe I might be just confused myself:

Since this GEP proposes encryption for the cluster-internal network hop from the gateway proxy to the backend, shouldn't the fields here be something like:

* `certificateRef` that is a reference to `Secret` of `type: kubernetes.io/tls`, where the secret includes keys `tls.crt` and `tls.key` which are the client credentials that gateway should use when establishing TLS connection to said backend service
* `caCertificateRef` that is a reference to `Secret` of `type: Opaque`, where the secret includes key `ca.crt` which is a CA bundle for validating the backend server certificate.

Sidenote: there is no common API / agreement about the CA certificate. For example, certificate-manager can store it in `type: kubernetes.io/tls` with name `ca.crt` even though `type: kubernetes.io/tls` does not recognize key with that name.

Reply:
I made some changes. Let me know what you think.